### PR TITLE
Implement explicit this type

### PR DESCRIPTION
**Pull request by dcodeIO**
It's already possible to declare an explicit this type as a `this:` argument, but the compiler doesn't yet make use of it. An explicit `this` type is especially useful to implement backing classes for basic values like `i32` that provide the respective `#toString()` functionality and as a building block for iterators which can possibly be represented by a single `i64` value without any allocation overhead. This PR aims at figuring this out.

### CODE REVIEW

File: tests/compiler/number.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import "allocator/arena";
+
+var a = 1;
+assert(a.toString() == "1");
+
+// TODO:
+// 1.toString()
```
Comment (MaxGraey) [inline tests/compiler/number.ts]: btw in js/ts this looks only like `1..toString()` but we have also integers so this should be legit in AS world

**Comment by dcodeIO** [comment]
Turns out that this opens up a new way to handle any kind of expression in the resolver, because with the backing basic classes, every expression now has a backing IR element. So, if I ever mentioned that something isn't possible because we are lacking a proper `resolveExpression`, that might now become obsolete.

File: std/assembly/number.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  itoa,
+  dtoa
+} from "./internal/number";
+
+abstract class Number<T> {
+  private constructor() {
+    unreachable();
+  }
+
+  toString(this: T): String {
```
Comment (MaxGraey) [inline std/assembly/number.ts]: I think need add `TODO` comment about optional `radix` param.

**Comment by dcodeIO** [comment]
With backing classes in place, I also think that we should move the respective static fields and methods to them, like `f64.EPSILON` (which looks more like Rust) should become `F64.ESPILON` etc., to mimic `Number` as close as possible.

File: std/assembly/number.ts (hunk 2 above)
Comment (dcodeIO) [reply std/assembly/number.ts]: Yep added a comment

File: tests/compiler/number.ts (hunk 1 above)
Comment (dcodeIO) [reply tests/compiler/number.ts]: Yeah, this is due to `1.` being parsed as a float, when followed by `toString` expects it to be an identifier, breaking the property access. Also happens in AS, but it appears that JS parsers have a specific syntax error there while AS emits `toString is not defined`.

**Comment by MaxGraey** [comment]
@dcodeIO definitely makes sense

**Comment by dcodeIO** [comment]
Merging so we can continue from here. Just note that `someString + someInt` coercion (implicitly calling `I32#toString(this: someInt)`) isn't implemented yet.

**Pull request merged**